File: gitlab_downloader/author_mapper.py
"""Author and committer mapping management for git migration."""

from __future__ import annotations

import json
import logging
import os
from pathlib import Path
from typing import Any

import yaml

from .models import AuthorMapping, CommitterMapping, MigrationConfig

logger = logging.getLogger(__name__)


class AuthorMapper:
    """Handles reading and writing of author/committer mappings to disk."""

    def __init__(self, config_path: str | Path):
        """Initialize AuthorMapper with a config file path.

        Args:
            config_path: Path to JSON or YAML config file
        """
        self.config_path = Path(config_path)

    def load_mappings(self) -> tuple[dict[str, AuthorMapping], dict[str, CommitterMapping]]:
        """Load author and committer mappings from config file.

        Returns:
            Tuple of (author_mappings, committer_mappings) dictionaries

        Raises:
            FileNotFoundError: If config file doesn't exist
            ValueError: If config format is invalid
        """
        if not self.config_path.exists():
            raise FileNotFoundError(f"Config file not found: {self.config_path}")

        file_content = self.config_path.read_text(encoding="utf-8")

        if self.config_path.suffix in {".yaml", ".yml"}:
            data = yaml.safe_load(file_content)
        elif self.config_path.suffix == ".json":
            data = json.loads(file_content)
        else:
            raise ValueError(f"Unsupported file format: {self.config_path.suffix}")

        if not isinstance(data, dict):
            raise ValueError("Config must be a dictionary")

        author_mappings = self._parse_author_mappings(data.get("author_mappings", {}))
        committer_mappings = self._parse_committer_mappings(data.get("committer_mappings", {}))

        return author_mappings, committer_mappings

    def save_mappings(
        self,
        author_mappings: dict[str, AuthorMapping],
        committer_mappings: dict[str, CommitterMapping],
    ) -> None:
        """Save author and committer mappings while preserving migration config.

        If the config file exists, preserves required migration fields.
        If the file doesn't exist, creates it with only the mappings.

        Args:
            author_mappings: Dictionary of author mappings
            committer_mappings: Dictionary of committer mappings

        Raises:
            ValueError: If config format is invalid
        """
        self.config_path.parent.mkdir(parents=True, exist_ok=True)

        data = {
            "author_mappings": self._serialize_mappings(author_mappings),
            "committer_mappings": self._serialize_mappings(committer_mappings),
        }

        # If file exists, preserve required migration fields
        if self.config_path.exists():
            file_content = self.config_path.read_text(encoding="utf-8")
            if self.config_path.suffix in {".yaml", ".yml"}:
                existing_data = yaml.safe_load(file_content)
            elif self.config_path.suffix == ".json":
                existing_data = json.loads(file_content)
            else:
                raise ValueError(f"Unsupported file format: {self.config_path.suffix}")

            if not isinstance(existing_data, dict):
                raise ValueError("Config must be a dictionary")

            # Preserve required migration fields if they exist
            required_fields = {"source_repos_path", "target_hosting_url", "target_token"}
            for field in required_fields:
                if field in existing_data:
                    data[field] = existing_data[field]

        if self.config_path.suffix in {".yaml", ".yml"}:
            content = yaml.dump(data, default_flow_style=False)
        elif self.config_path.suffix == ".json":
            content = json.dumps(data, indent=2)
        else:
            raise ValueError(f"Unsupported file format: {self.config_path.suffix}")

        # Write with restricted permissions (0o600) since config may contain tokens
        with os.fdopen(
            os.open(str(self.config_path), os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600),
            "w",
            encoding="utf-8",
        ) as f:
            f.write(content)

        logger.info(f"Mappings saved to {self.config_path}")

    def load_migration_config(self) -> MigrationConfig:
        """Load complete migration configuration from file.

        Returns:
            MigrationConfig instance

        Raises:
            FileNotFoundError: If config file doesn't exist
            ValueError: If config is invalid
        """
        if not self.config_path.exists():
            raise FileNotFoundError(f"Config file not found: {self.config_path}")

        file_content = self.config_path.read_text(encoding="utf-8")

        if self.config_path.suffix in {".yaml", ".yml"}:
            data = yaml.safe_load(file_content)
        elif self.config_path.suffix == ".json":
            data = json.loads(file_content)
        else:
            raise ValueError(f"Unsupported file format: {self.config_path.suffix}")

        if not isinstance(data, dict):
            raise ValueError("Config must be a dictionary")

        required_fields = {
            "source_repos_path",
            "target_hosting_url",
            "target_token",
        }
        missing = required_fields - set(data.keys())
        if missing:
            raise ValueError(f"Missing required config fields: {missing}")

        author_mappings = self._parse_author_mappings(data.get("author_mappings", {}))
        committer_mappings = self._parse_committer_mappings(data.get("committer_mappings", {}))

        return MigrationConfig(
            source_repos_path=data["source_repos_path"],
            target_hosting_url=data["target_hosting_url"],
            target_token=data["target_token"],
            author_mappings=author_mappings,
            committer_mappings=committer_mappings,
        )

    def save_migration_config(self, config: MigrationConfig) -> None:
        """Save complete migration configuration to file.

        Args:
            config: MigrationConfig instance to save
        """
        self.config_path.parent.mkdir(parents=True, exist_ok=True)

        data = {
            "source_repos_path": config.source_repos_path,
            "target_hosting_url": config.target_hosting_url,
            "target_token": config.target_token,
            "author_mappings": self._serialize_mappings(config.author_mappings),
            "committer_mappings": self._serialize_mappings(config.committer_mappings),
        }

        if self.config_path.suffix in {".yaml", ".yml"}:
            content = yaml.dump(data, default_flow_style=False)
        elif self.config_path.suffix == ".json":
            content = json.dumps(data, indent=2)
        else:
            raise ValueError(f"Unsupported file format: {self.config_path.suffix}")

        # Write with restricted permissions (0o600) since config contains tokens
        with os.fdopen(
            os.open(str(self.config_path), os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600),
            "w",
            encoding="utf-8",
        ) as f:
            f.write(content)

        logger.info(f"Migration config saved to {self.config_path}")

    @staticmethod
    def _parse_mappings(data: Any, cls: type, label: str) -> dict[str, Any]:
        """Parse mappings from dictionary using the given dataclass type.

        Args:
            data: Raw mapping dictionary from config file
            cls: Dataclass type to instantiate (AuthorMapping or CommitterMapping)
            label: Human-readable label for log messages
        """
        if not isinstance(data, dict):
            return {}

        mappings: dict[str, Any] = {}
        for key, value in data.items():
            if not isinstance(value, dict):
                continue

            try:
                original_name = value.get("original_name", "")
                original_email = value.get("original_email", "")
                new_name = value.get("new_name", "")
                new_email = value.get("new_email", "")

                if not original_name or not original_email:
                    logger.warning(
                        f"Skipping {label} mapping '{key}': "
                        "original_name and original_email are required"
                    )
                    continue

                mapping = cls(
                    original_name=original_name,
                    original_email=original_email,
                    new_name=new_name,
                    new_email=new_email,
                )
                mappings[key] = mapping
            except (KeyError, TypeError):
                logger.warning(f"Invalid {label} mapping for key: {key}")

        return mappings

    @staticmethod
    def _parse_author_mappings(data: Any) -> dict[str, AuthorMapping]:
        """Parse author mappings from dictionary."""
        return AuthorMapper._parse_mappings(data, AuthorMapping, "author")

    @staticmethod
    def _parse_committer_mappings(data: Any) -> dict[str, CommitterMapping]:
        """Parse committer mappings from dictionary."""
        return AuthorMapper._parse_mappings(data, CommitterMapping, "committer")

    @staticmethod
    def _serialize_mappings(
        mappings: dict[str, Any],
    ) -> dict[str, dict[str, str]]:
        """Convert mapping objects to serializable dictionaries."""
        result = {}
        for key, mapping in mappings.items():
            result[key] = {
                "original_name": mapping.original_name,
                "original_email": mapping.original_email,
                "new_name": mapping.new_name,
                "new_email": mapping.new_email,
            }
        return result

File: lib/author-mapper.js
import { readFile, writeFile, mkdir } from 'node:fs/promises';
import { access } from 'node:fs/promises';
import { dirname, extname, join } from 'node:path';
import { load as yamlParse, dump as yamlStringify } from 'js-yaml';

/**
 * @typedef {object} AuthorMapping
 * @property {string} original_name
 * @property {string} original_email
 * @property {string} new_name
 * @property {string} new_email
 */

/**
 * @typedef {object} CommitterMapping
 * @property {string} original_name
 * @property {string} original_email
 * @property {string} new_name
 * @property {string} new_email
 */

/**
 * @typedef {object} MigrationConfig
 * @property {string} source_repos_path
 * @property {string} target_hosting_url
 * @property {string} target_token
 * @property {Record<string, AuthorMapping>} author_mappings
 * @property {Record<string, CommitterMapping>} committer_mappings
 */

const MIGRATION_FIELDS = ['source_repos_path', 'target_hosting_url', 'target_token'];
const MAPPING_FIELDS = ['original_name', 'original_email', 'new_name', 'new_email'];

/**
 * Config filenames searched in order for auto-discovery.
 */
export const CONFIG_FILENAMES = [
  'migration_config.json',
  'migration_config.yaml',
  'migration_config.yml',
];

/**
 * Parse raw data into a mapping dictionary.
 *
 * @param {*} data - Raw mapping section from config file
 * @param {string} label - Mapping type label for warnings (e.g., "author", "committer")
 * @returns {Record<string, AuthorMapping|CommitterMapping>}
 */
function parseMappings(data, _label) {
  if (!data || typeof data !== 'object' || Array.isArray(data)) {
    return {};
  }
  const result = {};
  for (const [key, value] of Object.entries(data)) {
    if (!value || typeof value !== 'object' || Array.isArray(value)) {
      continue;
    }
    try {
      const original_name = String(value.original_name ?? '');
      const original_email = String(value.original_email ?? '');
      const new_name = String(value.new_name ?? '');
      const new_email = String(value.new_email ?? '');
      if (!original_name && !original_email) {
        continue;
      }
      result[key] = { original_name, original_email, new_name, new_email };
    } catch {
      // skip malformed entries
    }
  }
  return result;
}

/**
 * Serialize mapping objects into plain dictionaries for storage.
 *
 * @param {Record<string, AuthorMapping|CommitterMapping>} mappings
 * @returns {Record<string, object>}
 */
function serializeMappings(mappings) {
  const result = {};
  for (const [key, m] of Object.entries(mappings)) {
    result[key] = {
      original_name: m.original_name,
      original_email: m.original_email,
      new_name: m.new_name,
      new_email: m.new_email,
    };
  }
  return result;
}

/**
 * Detect file format from extension.
 *
 * @param {string} filePath
 * @returns {'json'|'yaml'}
 */
function detectFormat(filePath) {
  const ext = extname(filePath).toLowerCase();
  if (ext === '.yaml' || ext === '.yml') return 'yaml';
  if (ext === '.json') return 'json';
  throw new Error(`Unsupported config file format: ${ext}`);
}

/**
 * Parse file content based on format.
 *
 * @param {string} content - File content
 * @param {'json'|'yaml'} format
 * @returns {*}
 */
function parseContent(content, format) {
  if (format === 'json') return JSON.parse(content);
  return yamlParse(content);
}

/**
 * Serialize data to string based on format.
 *
 * @param {object} data
 * @param {'json'|'yaml'} format
 * @returns {string}
 */
function serializeContent(data, format) {
  if (format === 'json') return JSON.stringify(data, null, 2) + '\n';
  return yamlStringify(data, { lineWidth: -1 });
}

/**
 * Write a file with restricted permissions (0o600).
 *
 * @param {string} filePath
 * @param {string} content
 */
async function writeSecure(filePath, content) {
  await mkdir(dirname(filePath), { recursive: true });
  await writeFile(filePath, content, { encoding: 'utf-8', mode: 0o600 });
}

/**
 * Check if a file exists.
 *
 * @param {string} filePath
 * @returns {Promise<boolean>}
 */
async function fileExists(filePath) {
  try {
    await access(filePath);
    return true;
  } catch {
    return false;
  }
}

// ─── AuthorMapper ───────────────────────────────────────────────────

/**
 * Load author and committer mappings from a config file (JSON or YAML).
 *
 * @param {string} configPath - Path to JSON/YAML config file
 * @returns {Promise<{authorMappings: Record<string, AuthorMapping>, committerMappings: Record<string, CommitterMapping>}>}
 */
export async function loadMappings(configPath) {
  if (!(await fileExists(configPath))) {
    throw new Error(`Config file not found: ${configPath}`);
  }
  const format = detectFormat(configPath);
  const content = await readFile(configPath, 'utf-8');
  const data = parseContent(content, format);
  if (!data || typeof data !== 'object' || Array.isArray(data)) {
    throw new Error('Config file must contain an object');
  }
  const authorMappings = parseMappings(data.author_mappings, 'author');
  const committerMappings = parseMappings(data.committer_mappings, 'committer');
  return { authorMappings, committerMappings };
}

/**
 * Save author and committer mappings to a config file.
 * Preserves existing migration fields (source_repos_path, etc.) if the file exists.
 *
 * @param {string} configPath
 * @param {Record<string, AuthorMapping>} authorMappings
 * @param {Record<string, CommitterMapping>} committerMappings
 */
export async function saveMappings(configPath, authorMappings, committerMappings) {
  const format = detectFormat(configPath);
  const data = {
    author_mappings: serializeMappings(authorMappings),
    committer_mappings: serializeMappings(committerMappings),
  };

  // Preserve existing migration config fields if file exists
  if (await fileExists(configPath)) {
    try {
      const existingContent = await readFile(configPath, 'utf-8');
      const existingData = parseContent(existingContent, format);
      if (existingData && typeof existingData === 'object' && !Array.isArray(existingData)) {
        for (const field of MIGRATION_FIELDS) {
          if (existingData[field] !== undefined) {
            data[field] = existingData[field];
          }
        }
      }
    } catch {
      // If existing file can't be read/parsed, just write new data
    }
  }

  const serialized = serializeContent(data, format);
  await writeSecure(configPath, serialized);
}

// ─── MigrationConfig loading ────────────────────────────────────────

/**
 * Load a complete migration config from a file.
 *
 * @param {string} configPath
 * @returns {Promise<MigrationConfig>}
 */
export async function loadMigrationConfig(configPath) {
  if (!(await fileExists(configPath))) {
    throw new Error(`Config file not found: ${configPath}`);
  }
  const format = detectFormat(configPath);
  const content = await readFile(configPath, 'utf-8');
  const data = parseContent(content, format);
  if (!data || typeof data !== 'object' || Array.isArray(data)) {
    throw new Error('Config file must contain an object');
  }
  const missing = MIGRATION_FIELDS.filter((f) => !data[f]);
  if (missing.length > 0) {
    throw new Error(`Missing required fields: ${missing.join(', ')}`);
  }
  return {
    source_repos_path: data.source_repos_path,
    target_hosting_url: data.target_hosting_url,
    target_token: data.target_token,
    author_mappings: parseMappings(data.author_mappings, 'author'),
    committer_mappings: parseMappings(data.committer_mappings, 'committer'),
  };
}

/**
 * Save a complete migration config to a file.
 *
 * @param {string} configPath
 * @param {MigrationConfig} config
 */
export async function saveMigrationConfig(configPath, config) {
  const format = detectFormat(configPath);
  const data = {
    source_repos_path: config.source_repos_path,
    target_hosting_url: config.target_hosting_url,
    target_token: config.target_token,
    author_mappings: serializeMappings(config.author_mappings),
    committer_mappings: serializeMappings(config.committer_mappings),
  };
  const serialized = serializeContent(data, format);
  await writeSecure(configPath, serialized);
}

// ─── Config file discovery ──────────────────────────────────────────

/**
 * Discover and load a migration config from a repository directory.
 * Searches for CONFIG_FILENAMES in order and returns the first valid one found.
 *
 * @param {string} repoPath - Repository directory path
 * @returns {Promise<MigrationConfig|null>} - Null if no config file found
 */
export async function discoverConfig(repoPath) {
  for (const filename of CONFIG_FILENAMES) {
    const filePath = join(repoPath, filename);
    if (await fileExists(filePath)) {
      return loadMigrationConfig(filePath);
    }
  }
  return null;
}

/**
 * Save a migration config to a repository directory.
 *
 * @param {string} repoPath
 * @param {MigrationConfig} config
 * @param {'json'|'yaml'} [format='json']
 */
export async function saveConfigToRepo(repoPath, config, format = 'json') {
  if (format !== 'json' && format !== 'yaml') {
    throw new Error(`Unsupported format: ${format}. Must be "json" or "yaml"`);
  }
  const filename = `migration_config.${format}`;
  const filePath = join(repoPath, filename);
  await saveMigrationConfig(filePath, config);
}

/**
 * Validate a migration config data object.
 *
 * @param {object} data
 * @returns {object} - The validated data (unchanged)
 */
export function validateConfigData(data) {
  if (!data || typeof data !== 'object' || Array.isArray(data)) {
    throw new Error('Config data must be an object');
  }
  const missing = MIGRATION_FIELDS.filter((f) => !data[f]);
  if (missing.length > 0) {
    throw new Error(`Missing required fields: ${missing.join(', ')}`);
  }
  for (const section of ['author_mappings', 'committer_mappings']) {
    if (data[section]) {
      if (typeof data[section] !== 'object' || Array.isArray(data[section])) {
        throw new Error(`${section} must be an object`);
      }
      for (const [key, value] of Object.entries(data[section])) {
        if (typeof value !== 'object' || Array.isArray(value)) {
          throw new Error(`${section}.${key} must be an object`);
        }
        for (const field of MAPPING_FIELDS) {
          if (value[field] === undefined) {
            throw new Error(`${section}.${key} is missing field: ${field}`);
          }
        }
      }
    }
  }
  return data;
}
